pandas/_libs/src/inference.pyx

```diff
@@ -606,9 +611,16 @@ cpdef bint is_bool_array(ndarray values):
         if n == 0:
             return False
 
-        for i in range(n):
-            if not util.is_bool_object(objbuf[i]):
-                return False
+        if skipna:
+            for i in range(n):
+                val = objbuf[i]
+                if not util._checknull(val) and not util.is_bool_object(val):
+                    return False
```

Comment:
I feel like this logic is refactor-able.  It seems very duplicative across all of the functions.  Not a big deal, but just an observation ATM.